lib/elixir/test/elixir/kernel/diagnostics_test.exs

```diff
@@ -0,0 +1,181 @@
+      """
+
+      output =
+        capture_raise(
+          """
+          1 +
+          """,
+          TokenMissingError
+        )
+
+      assert output == expected
+    end
+
+    test "TokenMissingError (no snippet)" do
+      expected = """
+      ** (TokenMissingError) token missing on nofile:2:1:
+          ┌─ error: nofile:2:1
+          missing terminator: end (for "fn" starting at line 1)
+      """
+
+      output =
+        capture_raise(
+          """
+          fn a
+          """,
+          TokenMissingError
+        )
+
+      assert output == expected
+    end
+
+    test "keeps trailing whitespace if under threshold" do
+      expected = """
+      ** (SyntaxError) invalid syntax found on nofile:1:23:
+         ┌─ error: nofile:1:23
+         │
+       1 │                   a + 😎
+         │                       ^
+         │
+         unexpected token: "😎" (column 23, code point U+****)
+      """
+
+      output =
+        capture_raise(
+          """
+                            a + 😎
+          """,
+          SyntaxError
+        )
+
+      assert output == expected
+    end
+
+    test "limits trailing whitespace if too many" do
+      expected = """
+      ** (SyntaxError) invalid syntax found on nofile:1:43:
+         ┌─ error: nofile:1:43
+         │
+       1 │ ...                   a + 😎
+         │                           ^
+         │
+         unexpected token: "😎" (column 43, code point U+****)
+      """
+
+      output =
+        capture_raise(
+          """
+                                                a + 😎
+          """,
+          SyntaxError
+        )
+
+      assert output == expected
+    end
+
+    test "shows stacktrace if present" do
+      fake_stacktrace = [
+        {:fake, :fun, 3, [file: "nofile", line: 10]},
+        {:real, :fun, 2, [file: "nofile", line: 10]}
+      ]
+
+      expected = """
+      ** (TokenMissingError) token missing on nofile:1:4:
+         ┌─ error: nofile:1:4
+         │
+       1 │ 1 -
+         │    ^
+         │
+         syntax error: expression is incomplete
+
+          nofile:10: :fake.fun/3
+          nofile:10: :real.fun/2
+      """
+
+      output =
+        capture_raise(
+          """
+          1 -
+          """,
+          TokenMissingError,
+          fake_stacktrace
+        )
+
+      assert output == expected
+    end
+
+    test "handles unicode" do
+      source = read_fixture("unicode_error._ex")
+      output = capture_raise(source, SyntaxError)
+
+      assert output =~ "😎"
+    after
+      purge(Sample)
+    end
+  end
+
+  defp capture_raise(source, exception, mock_stacktrace \\ []) do
+    e =
+      assert_raise exception, fn ->
+        ast = Code.string_to_quoted!(source, columns: true)
+        Code.eval_quoted(ast)
+      end
+
+    Exception.format(:error, e, mock_stacktrace)
+  end
+
+  defp read_fixture(name) do
+    fixture = "diagnostics/" <> name
+
+    fixture
+    |> PathHelpers.fixture_path()
```

Comment:
Let's not use fixtures for this and instead have the source code defined in the file. Having to access another file actually makes writing/debugging tests harder. :)

Reply:
I think we'll need fixtures here since we need to have a concrete file in disk where the compiler can read and report the errors when we're testing. This is more apparent on the diagnostics tests, since we don't have snippets.

I've added the `unicode.ex` fixture here because it will also be reused in the diagnostics PR to check that it handles unicode lines correctly.